<div style=" background-color: RGB(100,100,100);" >
<h1 style="margin: auto; padding: 20px 0; color:#fff; text-align: center">PROJET 6 - DATA ANALYST</h1>
<h2 style="margin: auto; padding: 20px 0; color:#fff; text-align: center">Analyser les ventes d'une librairies avec Python
</h2>
</div>

<div style="background-color: RGB(140,140,140);" >
<h2 style="margin: auto; padding: 20px; color:#fff; ">Etape de nettoyage des fichiers</h2>
</div> 

In [113]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
from scipy import stats
from scipy.stats import chi2_contingency

In [114]:
customers = pd.read_csv("customers.csv")
products = pd.read_csv("products.csv")
transactions = pd.read_csv("transactions.csv")

In [115]:
customers1=customers.copy()
products1=products.copy()
transactions1=transactions.copy()

In [116]:
display (customers1)
display (products1)
display (transactions1)

client_id sex  birth
0       c_4410   f   1967
1       c_7839   f   1975
2       c_1699   f   1984
3       c_5961   f   1962
4       c_5320   m   1943
...        ...  ..    ...
8618    c_7920   m   1956
8619    c_7403   f   1970
8620    c_5119   m   1974
8621    c_5643   f   1968
8622      c_84   f   1982

[8623 rows x 3 columns]

id_prod   price  categ
0     0_1421   19.99      0
1     0_1368    5.13      0
2      0_731   17.99      0
3      1_587    4.99      1
4     0_1507    3.99      0
...      ...     ...    ...
3282    2_23  115.99      2
3283   0_146   17.14      0
3284   0_802   11.22      0
3285   1_140   38.56      1
3286  0_1920   25.16      0

[3287 rows x 3 columns]

id_prod                        date session_id client_id
0       0_1518  2022-05-20 13:21:29.043970   s_211425     c_103
1        1_251  2022-02-02 07:55:19.149409   s_158752    c_8534
2       0_1277  2022-06-18 15:44:33.155329   s_225667    c_6714
3        2_209  2021-06-24 04:19:29.835891    s_52962    c_6941
4       0_1509  2023-01-11 08:22:08.194479   s_325227    c_4232
...        ...                         ...        ...       ...
679527  0_1551  2022-01-15 13:05:06.246925   s_150195    c_8489
679528   1_639  2022-03-19 16:03:23.429229   s_181434    c_4370
679529  0_1425  2022-12-20 04:33:37.584749   s_314704     c_304
679530  0_1994  2021-07-16 20:36:35.350579    s_63204    c_2227
679531   1_523  2022-09-28 01:12:01.973763   s_274568    c_3873

[679532 rows x 4 columns]

# Vérifications des types de données

In [117]:
# Fichier customers1
df = pd.DataFrame(customers1)
types_de_donnees = df.dtypes
print(types_de_donnees)

client_id    object
sex          object
birth         int64
dtype: object


In [118]:
# Fichier products1
df = pd.DataFrame(products1)
types_de_donnees = df.dtypes
print(types_de_donnees)

id_prod     object
price      float64
categ        int64
dtype: object


In [119]:
# Fichier transactions1
df = pd.DataFrame(transactions1)
types_de_donnees = df.dtypes
print(types_de_donnees)

id_prod       object
date          object
session_id    object
client_id     object
dtype: object


# Vérifications des doublons 

In [120]:
# Fichier customers1
df = pd.DataFrame (customers1)
lignes_doubles= df[df.duplicated()]
display (lignes_doubles)

Empty DataFrame
Columns: [client_id, sex, birth]
Index: []

In [121]:
# Fichier products1
df = pd.DataFrame(products1)
lignes_doubles = df[df.duplicated()]
display (lignes_doubles)

Empty DataFrame
Columns: [id_prod, price, categ]
Index: []

In [122]:
# Fichier transactions1
df = pd.DataFrame(transactions1)
lignes_doubles = df[df.duplicated()]
display(lignes_doubles)

id_prod                             date session_id client_id
27778      T_0  test_2021-03-01 02:30:02.237437        s_0      ct_1
52424      T_0  test_2021-03-01 02:30:02.237419        s_0      ct_0
96687      T_0  test_2021-03-01 02:30:02.237412        s_0      ct_1
130188     T_0  test_2021-03-01 02:30:02.237419        s_0      ct_0
139339     T_0  test_2021-03-01 02:30:02.237443        s_0      ct_1
...        ...                              ...        ...       ...
653098     T_0  test_2021-03-01 02:30:02.237432        s_0      ct_0
657830     T_0  test_2021-03-01 02:30:02.237417        s_0      ct_0
662081     T_0  test_2021-03-01 02:30:02.237427        s_0      ct_1
671647     T_0  test_2021-03-01 02:30:02.237424        s_0      ct_1
679180     T_0  test_2021-03-01 02:30:02.237425        s_0      ct_1

[126 rows x 4 columns]

126 lignes sont en doublons dans le fichier transactions cependant la date n'a pas un code identique. 
Nous pouvons voir que ce sont des lignes de Test donc pas de sessions clients réelles

In [123]:
transactions1[transactions1["id_prod"]=="T_0"]

id_prod                             date session_id client_id
3019       T_0  test_2021-03-01 02:30:02.237419        s_0      ct_0
5138       T_0  test_2021-03-01 02:30:02.237425        s_0      ct_0
9668       T_0  test_2021-03-01 02:30:02.237437        s_0      ct_1
10728      T_0  test_2021-03-01 02:30:02.237436        s_0      ct_0
15292      T_0  test_2021-03-01 02:30:02.237430        s_0      ct_0
...        ...                              ...        ...       ...
657830     T_0  test_2021-03-01 02:30:02.237417        s_0      ct_0
662081     T_0  test_2021-03-01 02:30:02.237427        s_0      ct_1
670680     T_0  test_2021-03-01 02:30:02.237449        s_0      ct_1
671647     T_0  test_2021-03-01 02:30:02.237424        s_0      ct_1
679180     T_0  test_2021-03-01 02:30:02.237425        s_0      ct_1

[200 rows x 4 columns]

Nous avons 200 lignes de notre dataframe contenant un id_prod "test". 
Je décide de supprimer ces lignes de tests avant de travailler sur mes données

In [124]:
transactions1 = transactions1.drop(transactions1[transactions1["id_prod"] == "T_0"].index)
display (transactions1)

id_prod                        date session_id client_id
0       0_1518  2022-05-20 13:21:29.043970   s_211425     c_103
1        1_251  2022-02-02 07:55:19.149409   s_158752    c_8534
2       0_1277  2022-06-18 15:44:33.155329   s_225667    c_6714
3        2_209  2021-06-24 04:19:29.835891    s_52962    c_6941
4       0_1509  2023-01-11 08:22:08.194479   s_325227    c_4232
...        ...                         ...        ...       ...
679527  0_1551  2022-01-15 13:05:06.246925   s_150195    c_8489
679528   1_639  2022-03-19 16:03:23.429229   s_181434    c_4370
679529  0_1425  2022-12-20 04:33:37.584749   s_314704     c_304
679530  0_1994  2021-07-16 20:36:35.350579    s_63204    c_2227
679531   1_523  2022-09-28 01:12:01.973763   s_274568    c_3873

[679332 rows x 4 columns]

In [125]:
if any(transactions1["id_prod"].str.startswith("T")):
    print("Il reste des lignes avec un id_prod commençant par 'T'.")
else:
    print("Il ne reste pas de lignes avec un id_prod commençant par 'T'.")

Il ne reste pas de lignes avec un id_prod commençant par 'T'.


In [126]:
df = pd.DataFrame(transactions1)
lignes_doubles = df[df.duplicated()]
display(lignes_doubles)

Empty DataFrame
Columns: [id_prod, date, session_id, client_id]
Index: []

# Vérifications des valeurs manquantes

In [127]:
# Fichier customers1
df = pd.DataFrame(customers1)

# Vérifier s'il y a des valeurs manquantes dans le DataFrame
if df.isna().any().any():
    print("Le DataFrame contient des valeurs manquantes.")
else:
    print("Le DataFrame ne contient pas de valeurs manquantes.")

Le DataFrame ne contient pas de valeurs manquantes.


In [128]:
# Fichier customers1
df = pd.DataFrame(customers1)

# Vérifier s'il y a des valeurs manquantes dans le DataFrame
if df.isnull().any().any():
    print("Le DataFrame contient des valeurs nulles.")
else:
    print("Le DataFrame ne contient pas de valeurs nulles.")

Le DataFrame ne contient pas de valeurs nulles.


In [129]:
# Fichier products1
df = pd.DataFrame(products1)

# Vérifier s'il y a des valeurs manquantes dans le DataFrame
if df.isna().any().any():
    print("Le DataFrame contient des valeurs manquantes.")
else:
    print("Le DataFrame ne contient pas de valeurs manquantes.")

Le DataFrame ne contient pas de valeurs manquantes.


In [130]:
# Fichier customers1
df = pd.DataFrame(products1)

# Vérifier s'il y a des valeurs manquantes dans le DataFrame
if df.isnull().any().any():
    print("Le DataFrame contient des valeurs nulles.")
else:
    print("Le DataFrame ne contient pas de valeurs nulles.")

Le DataFrame ne contient pas de valeurs nulles.


In [131]:
# Fichier transactions1
df = pd.DataFrame(transactions1)

# Vérifier s'il y a des valeurs manquantes dans le DataFrame
if df.isna().any().any():
    print("Le DataFrame contient des valeurs manquantes.")
else:
    print("Le DataFrame ne contient pas de valeurs manquantes.")

Le DataFrame ne contient pas de valeurs manquantes.


In [132]:
# Fichier transactions1
df = pd.DataFrame(transactions1)

# Vérifier s'il y a des valeurs manquantes dans le DataFrame
if df.isnull().any().any():
    print("Le DataFrame contient des valeurs nulles.")
else:
    print("Le DataFrame ne contient pas de valeurs nulles.")

Le DataFrame ne contient pas de valeurs nulles.


# Ajout de colonne "age"dans mon dataframe customers

In [133]:
# Fichier customers1
customers1["age"] = 2023 - customers1["birth"]
display (customers1)

client_id sex  birth  age
0       c_4410   f   1967   56
1       c_7839   f   1975   48
2       c_1699   f   1984   39
3       c_5961   f   1962   61
4       c_5320   m   1943   80
...        ...  ..    ...  ...
8618    c_7920   m   1956   67
8619    c_7403   f   1970   53
8620    c_5119   m   1974   49
8621    c_5643   f   1968   55
8622      c_84   f   1982   41

[8623 rows x 4 columns]

# Vérification complémentaires des types de données et des valeurs étranges

In [134]:
# Fichier customers1
birth_max = customers1["birth"].max()
print ("L'année de naissance la plus récente est :",birth_max)
birth_min = customers1["birth"].min()
print ("L'année de naissance la plus ancienne est :",birth_min)

L'année de naissance la plus récente est : 2004
L'année de naissance la plus ancienne est : 1929


In [135]:
# Fichier customers1
age_max = customers1["age"].max()
print ("L'age le plus élévé dans nos clients est :",age_max)
age_min = customers1["age"].min()
print ("L'age le plus petit dans nos clients est :",age_min)
age_moyen = customers1["age"].mean()
print("La moyenne d'âge dans nos clients est :", round(age_moyen, 0))
age_mode = customers1["age"].mode().iloc[0]
print("La mode d'âge dans nos clients est :", age_mode)

L'age le plus élévé dans nos clients est : 94
L'age le plus petit dans nos clients est : 19
La moyenne d'âge dans nos clients est : 45.0
La mode d'âge dans nos clients est : 19


In [136]:
sex_dtype= customers1["client_id"].dtype 

# Vérification si le type est différent de object
if sex_dtype != "object":
    print("La colonne 'client_id' contient des valeurs qui ne sont pas de type object.")
else:
    print("Toutes les valeurs de la colonne 'client_id' sont de type object.")


Toutes les valeurs de la colonne 'client_id' sont de type object.


In [137]:
sex_dtype = customers1["sex"].dtype 

# Vérification si le type est différent de object
if sex_dtype != "object":
    print("La colonne 'sex' contient des valeurs qui ne sont pas de type object.")
else:
    print("Toutes les valeurs de la colonne 'sex' sont de type object.")


Toutes les valeurs de la colonne 'sex' sont de type object.


In [138]:
birth_dtype = customers1["birth"].dtype

# Vérification si le type est différent de int64
if birth_dtype != "int64":
    print("La colonne 'birth' contient des valeurs qui ne sont pas de type int64.")
else:
    print("Toutes les valeurs de la colonne 'birth' sont de type int64.")


Toutes les valeurs de la colonne 'birth' sont de type int64.


In [139]:
age_type = customers1["age"].dtype

# Vérification si le type est différent de int64
if age_type != "int64":
    print("La colonne 'age' contient des valeurs qui ne sont pas de type int64.")
else:
    print("Toutes les valeurs de la colonne 'age' sont de type int64.")


Toutes les valeurs de la colonne 'age' sont de type int64.


# ______________________________________________________________

In [140]:
# Fichier products1
max_price = products1 ["price"].max()
print("Max pour le price est :", max_price)
min_price = products1 ["price"].min()
print ("Min pour le price est :", min_price)
moyenne_price = products1 ["price"].mean()
print ("Moyenne pour le price est :", moyenne_price)

Max pour le price est : 300.0
Min pour le price est : -1.0
Moyenne pour le price est : 21.85664131426833


Je détecte un prix négatif. De plus, le prix max est, quant à lui, très éloigné de notre moyenne de prix.

In [141]:
# Fichier products1
trie_produit = products1.sort_values(by=["price"], ascending=False)
print(trie_produit)

     id_prod   price  categ
947      2_2  300.00      2
724     2_76  254.44      2
394    2_158  247.22      2
1436   2_167  236.99      2
2779    2_30  233.54      2
...      ...     ...    ...
1211  0_1844    0.77      0
370    0_120    0.66      0
2272   0_528    0.62      0
2355   0_202    0.62      0
731      T_0   -1.00      0

[3287 rows x 3 columns]


In [142]:
products1[products1["id_prod"]=="T_0"]

id_prod  price  categ
731     T_0   -1.0      0

Le prix négatif est donc une session test. Je décide de supprimer cette ligne

In [143]:
#suppresion de la ligne Test "T_0" avec un prix négatif
products1 = products1.drop(products1[products1["id_prod"] == "T_0"].index)

In [144]:
trie_produit = products1.sort_values(by=["price"], ascending=False)
print(trie_produit)

     id_prod   price  categ
947      2_2  300.00      2
724     2_76  254.44      2
394    2_158  247.22      2
1436   2_167  236.99      2
2779    2_30  233.54      2
...      ...     ...    ...
1530  0_1620    0.80      0
1211  0_1844    0.77      0
370    0_120    0.66      0
2272   0_528    0.62      0
2355   0_202    0.62      0

[3286 rows x 3 columns]


In [145]:
max_categ = products1["categ"].max()
print("Max de la colonne 'categ' :", max_categ)

Max de la colonne 'categ' : 2


In [146]:
min_categ = products1["categ"].min()
print("Min de la colonne 'categ' :", min_categ)

Min de la colonne 'categ' : 0


In [147]:
mode_categ = products1["categ"].mode().iloc[0]
print("Mode de la colonne 'categ' :", mode_categ)

Mode de la colonne 'categ' : 0


In [148]:
categ_dtype = products1["categ"].dtype

# Vérification si le type est différent de int64
if categ_dtype != 'int64':
    print("La colonne 'categ' contient des valeurs qui ne sont pas de type int64.")
else:
    print("Toutes les valeurs de la colonne 'categ' sont de type int64.")

Toutes les valeurs de la colonne 'categ' sont de type int64.


In [149]:
id_prod_dtype = products1["id_prod"].dtype

# Vérification si le type est différent de object
if id_prod_dtype != "object":
    print("La colonne 'id_prod' contient des valeurs qui ne sont pas de type object.")
else:
    print("Toutes les valeurs de la colonne 'id_prod' sont de type object.")

Toutes les valeurs de la colonne 'id_prod' sont de type object.


In [150]:
price_dtype = products1["price"].dtype

# Vérification si le type est différent de float64
if price_dtype != "float64":
    print("La colonne 'price' contient des valeurs qui ne sont pas de type float.")
else:
    print("Toutes les valeurs de la colonne 'price' sont de type float.")


Toutes les valeurs de la colonne 'price' sont de type float.


# _______________________________________________________________

In [151]:
# Fichier transactions1
id_prod_dtype = transactions["id_prod"].dtype

# Vérification si le type est différent de float64
if id_prod_dtype != "object":
    print("La colonne 'id_prod' contient des valeurs qui ne sont pas de type object.")
else:
    print("Toutes les valeurs de la colonne 'id_prod' sont de type object.")

Toutes les valeurs de la colonne 'id_prod' sont de type object.


In [152]:
date_dtype = transactions1["date"].dtype
print ("Le format de donnée pour la colonne 'date' est :", date_dtype)

Le format de donnée pour la colonne 'date' est : object


In [153]:
# Convertir la colonne "date" en type datetime
transactions1["date"] = pd.to_datetime(transactions1["date"])

# Formater les valeurs pour conserver uniquement jusqu'à la seconde (pas besoin des micro_secondes)
transactions1["date"] = transactions1["date"].dt.strftime("%Y-%m-%d %H:%M:%S")

# Afficher les 5 premières lignes pour vérifier le résultat
print(transactions1.head())

  id_prod                 date session_id client_id
0  0_1518  2022-05-20 13:21:29   s_211425     c_103
1   1_251  2022-02-02 07:55:19   s_158752    c_8534
2  0_1277  2022-06-18 15:44:33   s_225667    c_6714
3   2_209  2021-06-24 04:19:29    s_52962    c_6941
4  0_1509  2023-01-11 08:22:08   s_325227    c_4232


In [154]:
# Convertir la colonne "date" en type datetime
transactions1["date"] = pd.to_datetime(transactions1["date"])

# Vérifier le type de données après conversion
date_dtype = transactions1["date"].dtype
print(date_dtype)

datetime64[ns]


In [161]:
date_min = transactions1["date"].min()
print("La date la plus ancienne est :", date_min)

La date la plus ancienne est : 2021-03-01 00:01:07


In [160]:
date_max = transactions1["date"].max()
print("La date la plus récente est :", date_max)

La date la plus récente est : 2023-02-28 23:58:30


In [157]:
session_id_dtype = transactions["session_id"].dtype

# Vérification si le type est différent de float64
if session_id_dtype != "object":
    print("La colonne 'session_id' contient des valeurs qui ne sont pas de type object.")
else:
    print("Toutes les valeurs de la colonne 'session_id' sont de type object.")

Toutes les valeurs de la colonne 'session_id' sont de type object.


In [158]:
client_id_dtype = transactions["client_id"].dtype

# Vérification si le type est différent de float64
if client_id_dtype != "object":
    print("La colonne 'client_id' contient des valeurs qui ne sont pas de type object.")
else:
    print("Toutes les valeurs de la colonne 'client_id' sont de type object.")

Toutes les valeurs de la colonne 'client_id' sont de type object.


In [159]:
trie_transactions1=transactions1.sort_values(by="date", ascending=True)
display(trie_transactions1)

id_prod                date session_id client_id
480642  0_1259 2021-03-01 00:01:07        s_1     c_329
428477  0_1390 2021-03-01 00:02:26        s_2     c_664
462702  0_1352 2021-03-01 00:02:38        s_3     c_580
488300  0_1458 2021-03-01 00:04:54        s_4    c_7912
439073  0_1358 2021-03-01 00:05:18        s_5    c_2033
...        ...                 ...        ...       ...
184649   1_508 2023-02-28 23:49:03   s_348444    c_3573
588183    2_37 2023-02-28 23:51:29   s_348445      c_50
130172   1_695 2023-02-28 23:53:18   s_348446     c_488
550455  0_1547 2023-02-28 23:58:00   s_348447    c_4848
84274   0_1398 2023-02-28 23:58:30   s_348435    c_3575

[679332 rows x 4 columns]